<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Assembling reads with Velvet &mdash; Metagenomics Workshop SciLifeLab 1.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="Metagenomics Workshop SciLifeLab 1.0 documentation" href="../index.html" />
    <link rel="up" title="Metagenomic Assembly Workshop" href="index.html" />
    <link rel="next" title="Mapping reads back to the assembly" href="map.html" />
    <link rel="prev" title="Quality trimming Illumina paired-end reads" href="qtrim.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="map.html" title="Mapping reads back to the assembly"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="qtrim.html" title="Quality trimming Illumina paired-end reads"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Metagenomics Workshop SciLifeLab 1.0 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Metagenomic Assembly Workshop</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="assembling-reads-with-velvet">
<h1>Assembling reads with Velvet<a class="headerlink" href="#assembling-reads-with-velvet" title="Permalink to this headline">¶</a></h1>
<p>In this exercise we will learn how to perform an assembly with Velvet. Velvet
takes your reads as input and turns them into contigs. It consists of two
steps. In the first step, <tt class="docutils literal"><span class="pre">velveth</span></tt>, the de Bruijn graph is created.
Afterwards the graph is traversed and contigs are created with <tt class="docutils literal"><span class="pre">velvetg</span></tt>.
When constructing the de Bruijn graph, a <em>kmer</em> has to be specified. Reads are
cut up into pieces of length <em>k</em>, each representing a node in the graph, edges
represent an overlap (some de Bruijn graph assemblers do this differently, but
the idea is the same). The advantage of using kmer overlap instead of read
overlap is that the computational requirements grow with the number of unique
kmers instead of unique reads. A more detailled explanation can be found at
<a class="reference external" href="http://www.nature.com/nbt/journal/v29/n11/full/nbt.2023.html">http://www.nature.com/nbt/journal/v29/n11/full/nbt.2023.html</a>.</p>
<div class="section" id="pick-a-kmer">
<h2>Pick a kmer<a class="headerlink" href="#pick-a-kmer" title="Permalink to this headline">¶</a></h2>
<p>Please work in pairs for this assignment. Every group can select a kmer of
their likings - pick a random one if you haven&#8217;t developed a preference yet.
Write you and your partner&#8217;s name down at a kmer on the
Google <a class="reference external" href="https://docs.google.com/spreadsheet/ccc?key=0AvduvUOYAB-_dGtrcV9hRUZxLWxreE5ENGs2OXRnM3c&amp;usp=drive_web#gid=3.">doc</a> for this workshop.</p>
</div>
<div class="section" id="velveth">
<h2>velveth<a class="headerlink" href="#velveth" title="Permalink to this headline">¶</a></h2>
<p>Create the graph data structure with <tt class="docutils literal"><span class="pre">velveth</span></tt>. Again like we did with
<tt class="docutils literal"><span class="pre">sickle</span></tt>, first create a directory with symbolic links to the pairs that you
want to use:</p>
<div class="highlight-python"><pre>mkdir -p ~/asm-workshop/velvet
cd ~/asm-workshop/velvet
ln -s ../sickle/qtrim1.fastq pair1.fastq
ln -s ../sickle/qtrim2.fastq pair2.fastq</pre>
</div>
<p>The reads need to be interleaved for <tt class="docutils literal"><span class="pre">velveth</span></tt>:</p>
<div class="highlight-python"><pre>shuffleSequences_fastq.pl pair1.fastq pair2.fastq pair.fastq</pre>
</div>
<p>Run velveth over the kmer you picked (21 in this example):</p>
<div class="highlight-python"><pre>velveth out_21 21 -fastq -shortPaired pair.fastq</pre>
</div>
<p>Check what directories have been created:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ls</span>
</pre></div>
</div>
</div>
<div class="section" id="velvetg">
<h2>velvetg<a class="headerlink" href="#velvetg" title="Permalink to this headline">¶</a></h2>
<p>To get the actual contigs you will have to run <tt class="docutils literal"><span class="pre">velvetg</span></tt> on the created
graph. You can vary options such expected coverage and the coverage cut-off if
you want, but we do not do that in this tutorial. We only choose not to do
scaffolding:</p>
<div class="highlight-python"><pre>velvetg out_21 -scaffolding no</pre>
</div>
</div>
<div class="section" id="assemstats">
<h2>assemstats<a class="headerlink" href="#assemstats" title="Permalink to this headline">¶</a></h2>
<p>After the assembly one wants to look at the length distributions of the
resulting assemblies. You can use the <tt class="docutils literal"><span class="pre">assemstats</span></tt> script for that:</p>
<div class="highlight-python"><pre>assemstats 100 out_*/contigs.fa</pre>
</div>
<p>Try to find-out what each of the stats represent by varying the cut-off. One of
the most often used statistics in assembly length distribution comparisons is
the <em>N50 length</em>, a weighted median, where you weight each contig by it&#8217;s
length. This way you assign more weight to larger contigs. Fifty procent of all
the bases in the assembly are contained in contigs shorter or equal to N50
length. Once you have gotten an idea of what it all the stats mean, it is time
to compare your results with the other attendees of this workshop. Generate the results and copy them to the <a class="reference external" href="https://docs.google.com/spreadsheet/ccc?key=0AvduvUOYAB-_dGtrcV9hRUZxLWxreE5ENGs2OXRnM3c&amp;usp=drive_web#gid=3.">doc</a>:</p>
<div class="highlight-python"><pre>assemstats 100 out_*/contigs.fa</pre>
</div>
<p>Do the same for the cut-off at 1000 and add it to the <a class="reference external" href="https://docs.google.com/spreadsheet/ccc?key=0AvduvUOYAB-_dGtrcV9hRUZxLWxreE5ENGs2OXRnM3c&amp;usp=drive_web#gid=3.">doc</a>. Compare your kmer
against the others. If there are very little available yet, this would be an
ideal time to help out some other attendees or do the same exercise for a kmer
that has not been picked by somebody else yet. Please write down you and your
partners name again at the <a class="reference external" href="https://docs.google.com/spreadsheet/ccc?key=0AvduvUOYAB-_dGtrcV9hRUZxLWxreE5ENGs2OXRnM3c&amp;usp=drive_web#gid=3.">doc</a> in that case.</p>
<p>What are the important length statistics? Do we prefer sum over length? Should
it be a combination? Think of a formula that could indicate the best preferred
length distribution where you express the optimization function in terms of the
column names from the <a class="reference external" href="https://docs.google.com/spreadsheet/ccc?key=0AvduvUOYAB-_dGtrcV9hRUZxLWxreE5ENGs2OXRnM3c&amp;usp=drive_web#gid=3.">doc</a>. For instance only <tt class="docutils literal"><span class="pre">n50_len</span></tt> or <tt class="docutils literal"><span class="pre">sum</span> <span class="pre">*</span>
<span class="pre">n50_len</span></tt>.</p>
</div>
<div class="section" id="optional-exercise-ray">
<h2>(Optional exercise) Ray<a class="headerlink" href="#optional-exercise-ray" title="Permalink to this headline">¶</a></h2>
<p>Try to create an assembly with Ray over the same kmer. Ray is an assembler that
uses MPI to distribute the assembly over multiple cores and nodes. The latest
version of Ray was made to work well with metagenomics data as well:</p>
<div class="highlight-python"><pre>mkdir -p ~/asm-workshop/ray
cd ~/asm-workshop/ray
ln -s ../sickle/qtrim1.fastq pair1.fastq
ln -s ../sickle/qtrim2.fastq pair2.fastq
mpiexec -n 1 Ray -k 21 -p pair1.fastq pair2.fastq -o out_21</pre>
</div>
<p>Add the <tt class="docutils literal"><span class="pre">assemstats</span></tt> results to the <a class="reference external" href="https://docs.google.com/spreadsheet/ccc?key=0AvduvUOYAB-_dGtrcV9hRUZxLWxreE5ENGs2OXRnM3c&amp;usp=drive_web#gid=3.">doc</a> as you did for Velvet. There is a
separate tab for the Ray assemblies, compare the results with Velvet.</p>
</div>
<div class="section" id="optional-exercise-velvetoptimiser">
<h2>(Optional exercise) VelvetOptimiser<a class="headerlink" href="#optional-exercise-velvetoptimiser" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://github.com/Victorian-Bioinformatics-Consortium/VelvetOptimiser">VelvetOptimiser</a> is a script that runs Velvet multiple times and follows the
optimization function you give it. Use <a class="reference external" href="https://github.com/Victorian-Bioinformatics-Consortium/VelvetOptimiser">VelvetOptimiser</a> to find the assembly
that gets the best score for the optimization function you designed in <a class="reference internal" href="#assemstats">assemstats</a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Assembling reads with Velvet</a><ul>
<li><a class="reference internal" href="#pick-a-kmer">Pick a kmer</a></li>
<li><a class="reference internal" href="#velveth">velveth</a></li>
<li><a class="reference internal" href="#velvetg">velvetg</a></li>
<li><a class="reference internal" href="#assemstats">assemstats</a></li>
<li><a class="reference internal" href="#optional-exercise-ray">(Optional exercise) Ray</a></li>
<li><a class="reference internal" href="#optional-exercise-velvetoptimiser">(Optional exercise) VelvetOptimiser</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="qtrim.html"
                        title="previous chapter">Quality trimming Illumina paired-end reads</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="map.html"
                        title="next chapter">Mapping reads back to the assembly</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/assembly/assembly.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="map.html" title="Mapping reads back to the assembly"
             >next</a> |</li>
        <li class="right" >
          <a href="qtrim.html" title="Quality trimming Illumina paired-end reads"
             >previous</a> |</li>
        <li><a href="../index.html">Metagenomics Workshop SciLifeLab 1.0 documentation</a> &raquo;</li>
          <li><a href="index.html" >Metagenomic Assembly Workshop</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Johannes Alneberg, Johan Bengtsson-Palme, Ino de Bruijn, Luisa Hugerth, Mikael Huss, Thomas Svensson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b3.
    </div>
  </body>
</html>