<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Translating nucleotide sequences into amino acid sequences &mdash; Metagenomics Workshop SciLifeLab 1.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="Metagenomics Workshop SciLifeLab 1.0 documentation" href="../index.html" />
    <link rel="up" title="Metagenomic Annotation Workshop" href="index.html" />
    <link rel="next" title="Search amino acid sequences with HMMER against the Pfam database" href="annotation.html" />
    <link rel="prev" title="Checking required software" href="software.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="annotation.html" title="Search amino acid sequences with HMMER against the Pfam database"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="software.html" title="Checking required software"
             accesskey="P">previous</a> |</li>
        <li><a href="../index.html">Metagenomics Workshop SciLifeLab 1.0 documentation</a> &raquo;</li>
          <li><a href="index.html" accesskey="U">Metagenomic Annotation Workshop</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="translating-nucleotide-sequences-into-amino-acid-sequences">
<h1>Translating nucleotide sequences into amino acid sequences<a class="headerlink" href="#translating-nucleotide-sequences-into-amino-acid-sequences" title="Permalink to this headline">¶</a></h1>
<p>The first step before we can annotate the metagenomes with Pfam domains
using HMMER will be to translate the reads into amino acid sequences. This
is necessary because HMMER (still) does not translate nucleotide sequnces
into protein space on the fly (like,for example, BLAST). For completing
this task we will use <tt class="docutils literal"><span class="pre">transeq</span></tt>, part of the <a class="reference external" href="http://emboss.sourceforge.net">EMBOSS</a>
package.</p>
<div class="section" id="running-transeq-on-the-sequence-data-sets">
<h2>Running <tt class="docutils literal"><span class="pre">transeq</span></tt> on the sequence data sets<a class="headerlink" href="#running-transeq-on-the-sequence-data-sets" title="Permalink to this headline">¶</a></h2>
<p>To run <tt class="docutils literal"><span class="pre">transeq</span></tt>, take a look at its available options:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">transeq</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
<p>A few options are important in this context. First of all, we need to
supply an input file, using the (somewhat bulky) option <tt class="docutils literal"><span class="pre">-sequence</span></tt>.
Second, we also need to specify an output file, otherwise transeq will
simply write its output to the screen. This is specified using the
<tt class="docutils literal"><span class="pre">-outseq</span></tt> option.</p>
<p>However, if we just run <tt class="docutils literal"><span class="pre">transeq</span></tt> like this we will
run into two additional problems. First, <tt class="docutils literal"><span class="pre">transeq</span></tt> by default just
translate the reading frame beginning at the first base in the input sequnece,
and will ignore any bases in the reading frames beginning with base two
and three, as well as those on the reverse-complementary strand. Second,
the software will add stop characters in the form of asterixes <tt class="docutils literal"><span class="pre">*</span></tt> whenever
it encounters a stop codon. This will occasionally cause HMMER to choke, so we
want stop codons to instead be translated into X characters that HMMER can handle.
The following excerpt form the <a class="reference external" href="http://selab.janelia.org/people/eddys/blog/?p=424">HMMER creator&#8217;s blog</a>
on this subject is one of my personal all-time favorites in terms of computer
software documentation:</p>
<blockquote>
<div>There’s two ways people do six-frame translation. You can translate each
frame into separate ORFs, or you can translate the read into exactly six
“ORFs”, one per frame, with * chararacters marking stop codons. HMMER
prefers that you do the former. Technically, * chararacters aren’t legal
amino acid residue codes, and the author of HMMER3 is a pedantic nitpicker,
passive-aggressive, yet also a pragmatist: so while HMMER3 pragmatically
accepts * chararacters in input “protein” sequences just fine, it pedantically
relegates them to somewhat suboptimal status, and it passively-aggressively
figures that any suboptimal performance on *-containing ORFs is your own
fault for using *’s in the first place.</div></blockquote>
<p>To avoid making Sean Eddy angry and causing other problems for our HMMER runs,
we will use the <tt class="docutils literal"><span class="pre">-frame</span> <span class="pre">6</span></tt> option to <tt class="docutils literal"><span class="pre">transeq</span></tt> in order to get translations
of all six reading frames, and the <tt class="docutils literal"><span class="pre">-clean</span></tt> option to convert stop codons to X
instead of *.</p>
<p>That should give us the command:</p>
<div class="highlight-python"><div class="highlight"><pre>transeq -sequence &lt;input file&gt; -outseq &lt;output file&gt; -frame 6 -clean
</pre></div>
</div>
<p>Now run this command on all four input files that we have created links to.
When the command has finished for all files, we can move on to the actual
annotation.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Translating nucleotide sequences into amino acid sequences</a><ul>
<li><a class="reference internal" href="#running-transeq-on-the-sequence-data-sets">Running <tt class="docutils literal"><span class="pre">transeq</span></tt> on the sequence data sets</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="software.html"
                        title="previous chapter">Checking required software</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="annotation.html"
                        title="next chapter">Search amino acid sequences with HMMER against the Pfam database</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/annotation/translation.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="annotation.html" title="Search amino acid sequences with HMMER against the Pfam database"
             >next</a> |</li>
        <li class="right" >
          <a href="software.html" title="Checking required software"
             >previous</a> |</li>
        <li><a href="../index.html">Metagenomics Workshop SciLifeLab 1.0 documentation</a> &raquo;</li>
          <li><a href="index.html" >Metagenomic Annotation Workshop</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2014, Johannes Alneberg, Johan Bengtsson-Palme, Ino de Bruijn, Luisa Hugerth, Mikael Huss, Thomas Svensson.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
  </body>
</html>